File: v1/HLD/MessageQueues/code/MessageQueue.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <queue>
#include <mutex>
#include <vector>
#include <map>
#include <string>

const int NUM_MESSAGES = 20;
const int PROCESSING_MS = 50;

double processWithWorkers(int numWorkers, std::map<std::string, int>& processedBy) {
    std::queue<std::string> q;
    for (int i = 0; i < NUM_MESSAGES; i++) {
        q.push("message-" + std::to_string(i));
    }
    std::mutex queueMutex;
    std::mutex resultMutex;

    auto worker = [&](int workerId) {
        while (true) {
            std::string message;
            {
                std::lock_guard<std::mutex> lock(queueMutex);
                if (q.empty()) return;
                message = q.front();
                q.pop();
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(PROCESSING_MS));
            {
                std::lock_guard<std::mutex> lock(resultMutex);
                processedBy[message] = workerId;
            }
        }
    };

    auto start = std::chrono::steady_clock::now();
    std::vector<std::thread> threads;
    for (int w = 0; w < numWorkers; w++) {
        threads.emplace_back(worker, w);
    }
    for (auto& t : threads) t.join();
    return std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - start).count();
}

int main() {
    std::map<std::string, int> processedBy1;
    double elapsed1 = processWithWorkers(1, processedBy1);
    std::cout << "1 consumer:  " << NUM_MESSAGES << " messages in " << elapsed1 << " ms" << std::endl;

    std::map<std::string, int> processedBy3;
    double elapsed3 = processWithWorkers(3, processedBy3);
    std::map<int, int> counts;
    for (auto& kv : processedBy3) counts[kv.second]++;
    std::cout << "3 consumers: " << NUM_MESSAGES << " messages in " << elapsed3 << " ms" << std::endl;
    std::cout << "Messages per worker:";
    for (auto& kv : counts) std::cout << " worker-" << kv.first << "=" << kv.second;
    std::cout << std::endl;
    std::cout << "Speedup from 3 concurrent consumers: " << (elapsed1 / elapsed3) << "x" << std::endl;
    return 0;
}

File: v1/HLD/MessageQueues/code/MessageQueue.java
import java.util.*;
import java.util.concurrent.*;

public class MessageQueue {
    static final int NUM_MESSAGES = 20;
    static final int PROCESSING_MS = 50;

    static double processWithWorkers(int numWorkers, Map<String, Integer> processedBy) throws InterruptedException {
        ConcurrentLinkedQueue<String> queue = new ConcurrentLinkedQueue<>();
        for (int i = 0; i < NUM_MESSAGES; i++) {
            queue.add("message-" + i);
        }
        Map<String, Integer> results = new ConcurrentHashMap<>();

        long start = System.nanoTime();
        Thread[] threads = new Thread[numWorkers];
        for (int w = 0; w < numWorkers; w++) {
            final int workerId = w;
            threads[w] = new Thread(() -> {
                String message;
                while ((message = queue.poll()) != null) {
                    try {
                        Thread.sleep(PROCESSING_MS);
                    } catch (InterruptedException ignored) {
                    }
                    results.put(message, workerId);
                }
            });
            threads[w].start();
        }
        for (Thread t : threads) t.join();
        processedBy.putAll(results);
        return (System.nanoTime() - start) / 1_000_000.0;
    }

    public static void main(String[] args) throws InterruptedException {
        Map<String, Integer> processedBy1 = new HashMap<>();
        double elapsed1 = processWithWorkers(1, processedBy1);
        System.out.printf("1 consumer:  %d messages in %.1f ms%n", NUM_MESSAGES, elapsed1);

        Map<String, Integer> processedBy3 = new HashMap<>();
        double elapsed3 = processWithWorkers(3, processedBy3);
        Map<Integer, Integer> counts = new TreeMap<>();
        for (int workerId : processedBy3.values()) {
            counts.merge(workerId, 1, Integer::sum);
        }
        System.out.printf("3 consumers: %d messages in %.1f ms%n", NUM_MESSAGES, elapsed3);
        System.out.println("Messages per worker: " + counts);
        System.out.printf("Speedup from 3 concurrent consumers: %.2fx%n", elapsed1 / elapsed3);
    }
}
